File: web/engine/validity.js
// Whether a part can be built at all.
//
// CAD calls this a rebuild error. A dimension points at a variable somebody
// deleted, two equations refer to each other in a loop, or a value reaches
// zero and a feature collapses. These checks answer the same question
// without opening a CAD system.

import { resolve } from './expr.js';
import { buildGraph, cycles, dangling } from './graph.js';
import { activeFeatures, buildGrid, partBounds, resolvedParameters } from './part.js';
import { evaluateField } from './sdf.js';
import { labelRegions, occupancyFromField } from './voxels.js';

const POSITIVE_KEYS = new Set(['radius', 'height', 'tube']);

const issue = (severity, scope, where, message) => ({ severity, scope, where, message });

export function checkEquations(part) {
  const graph = buildGraph(part);
  const issues = [];

  for (const [where, missing] of dangling(graph)) {
    issues.push(issue('error', 'equation', where,
      `reads "${missing}", which the parameter table does not define`));
  }
  for (const loop of cycles(graph)) {
    issues.push(issue('error', 'equation', loop.join(' → '),
      'these parameters depend on each other in a loop'));
  }
  if (issues.length) return issues;

  let values;
  try {
    values = resolvedParameters(part);
  } catch (error) {
    return [issue('error', 'equation', 'parameters', error.message)];
  }
  for (const [name, value] of Object.entries(values)) {
    if (!Number.isFinite(value)) issues.push(issue('error', 'equation', name, `resolves to ${value}`));
  }
  return issues;
}

function number(raw, values) {
  try {
    return resolve(raw, values);
  } catch {
    return null;
  }
}

export function checkFeatures(part, values) {
  const issues = [];
  for (const feature of activeFeatures(part)) {
    for (const [key, raw] of Object.entries(feature)) {
      if (POSITIVE_KEYS.has(key)) {
        const value = number(raw, values);
        if (value !== null && value <= 0) {
          issues.push(issue('error', 'feature', `${feature.name || feature.id}.${key}`,
            `is ${value}, and a size has to be above zero`));
        }
      } else if (key === 'size' && Array.isArray(raw)) {
        raw.forEach((item, index) => {
          const value = number(item, values);
          if (value !== null && value <= 0) {
            const axis = index < 3 ? 'xyz'[index] : index;
            issues.push(issue('error', 'feature', `${feature.name || feature.id}.size.${axis}`,
              `is ${value}, and a size has to be above zero`));
          }
        });
      }
    }
  }
  return issues;
}

// Volume and body count without tessellating. Counting whole cells aliases
// badly on a face lying along the lattice, so each cell contributes the
// fraction the distance says it fills.
// A lattice of cell centres rather than cell corners.
//
// Each sample here stands for one whole cell, and the fraction of that cell
// the surface leaves filled is what the volume adds up. Sampling the corners
// instead counts a shell of half cells at full weight, which is a real bias
// and put this engine a tenth of a percent above the Python one. The command
// line tool measures on cell centres, and so does this.
function cellCentres(grid) {
  return {
    origin: grid.origin.map((value) => value + grid.pitch / 2),
    pitch: grid.pitch,
    dims: grid.dims.map((size) => Math.max(1, size - 1)),
  };
}

export function measureSolid(part, resolution = 26) {
  const params = resolvedParameters(part);
  const grid = cellCentres(buildGrid(partBounds(part, params), resolution));
  const field = evaluateField(activeFeatures(part), params, grid);
  const cell = grid.pitch ** 3;
  let filled = 0;
  for (let i = 0; i < field.length; i += 1) {
    const fraction = 0.5 - field[i] / grid.pitch;
    filled += fraction < 0 ? 0 : fraction > 1 ? 1 : fraction;
  }
  const occupancy = occupancyFromField(field);
  let inside = 0;
  for (let i = 0; i < occupancy.length; i += 1) inside += occupancy[i];
  if (!inside) return { volume: 0, bodies: 0 };
  return { volume: filled * cell, bodies: labelRegions(occupancy, grid.dims).count };
}

export function checkPart(part, { geometry = true, resolution = 24 } = {}) {
  const issues = checkEquations(part);
  if (issues.length) return issues;

  const values = resolvedParameters(part);
  issues.push(...checkFeatures(part, values));
  if (issues.length || !geometry) return issues;

  if (!activeFeatures(part).length) {
    return [issue('error', 'geometry', 'part', 'has no features left to build')];
  }

  let measured;
  try {
    measured = measureSolid(part, resolution);
  } catch (error) {
    return [issue('error', 'geometry', 'part', `cannot be evaluated: ${error.message}`)];
  }
  if (measured.volume <= 0) {
    issues.push(issue('error', 'geometry', 'part', 'builds to nothing at this size'));
  } else if (measured.bodies > 1) {
    issues.push(issue('warning', 'geometry', 'part',
      `has fallen into ${measured.bodies} separate bodies`));
  }
  return issues;
}

File: kerf/parametric/validity.py
"""Whether a part can be built at all.

CAD calls this a rebuild error. A dimension points at a variable somebody
deleted, two equations refer to each other in a loop, or a value goes to zero
and a feature collapses. The model stops regenerating, and the person who
finds out is usually not the person who caused it.

These checks answer the same question without opening a CAD system, which is
what lets kerf run them on a merge before the merge is allowed to land.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional

import numpy as np

from ..geometry.voxels import label_regions
from .expr import ExpressionError, resolve
from .graph import build_graph

# Fields that describe a size. A size of zero or less collapses the feature.
POSITIVE_KEYS = {"radius", "height", "tube"}


@dataclass
class ModelIssue:
    severity: str        # error or warning
    scope: str           # equation, feature, or geometry
    where: str
    message: str

    def describe(self) -> str:
        return f"{self.where}: {self.message}"


def check_equations(part) -> list[ModelIssue]:
    """Check the equations on their own, without building any geometry."""
    issues: list[ModelIssue] = []
    graph = build_graph(part)

    for where, missing in graph.dangling():
        issues.append(
            ModelIssue(
                "error", "equation", where,
                f"reads {missing!r}, which the parameter table does not define",
            )
        )

    for loop in graph.cycles():
        issues.append(
            ModelIssue(
                "error", "equation", " -> ".join(loop),
                "these parameters depend on each other in a loop",
            )
        )

    if issues:
        return issues                       # values cannot be trusted yet

    try:
        values = part.resolved_parameters()
    except ExpressionError as error:
        return [ModelIssue("error", "equation", "parameters", str(error))]

    for name, value in values.items():
        if not math.isfinite(value):
            issues.append(
                ModelIssue("error", "equation", name, f"resolves to {value}")
            )
    return issues


def check_features(part, values: dict[str, float]) -> list[ModelIssue]:
    """Check that every feature has dimensions it can be built from."""
    issues: list[ModelIssue] = []
    for feature in part.active_features():
        for key, raw in feature.params.items():
            if key in POSITIVE_KEYS:
                number = _number(raw, values)
                if number is not None and number <= 0:
                    issues.append(
                        ModelIssue(
                            "error", "feature", f"{feature.label()}.{key}",
                            f"is {number:g}, and a size has to be above zero",
                        )
                    )
            elif key == "size":
                for index, item in enumerate(raw if isinstance(raw, (list, tuple)) else []):
                    number = _number(item, values)
                    if number is not None and number <= 0:
                        axis = "xyz"[index] if index < 3 else str(index)
                        issues.append(
                            ModelIssue(
                                "error", "feature", f"{feature.label()}.size.{axis}",
                                f"is {number:g}, and a size has to be above zero",
                            )
                        )
    return issues


def _number(raw, values: dict[str, float]):
    try:
        return resolve(raw, values)
    except Exception:                        # noqa: BLE001
        return None


def measure_solid(part, resolution: int = 32) -> tuple[float, int]:
    """Volume and body count, taken from the field without tessellating.

    A sweep evaluates a part many times over, and the mesh is not needed to
    answer whether the part still exists and whether it is still in one piece.

    Counting cells whose centre is inside would be simpler and it aliases
    badly, because a flat face lying near a layer of centres flips that whole
    layer in or out. The distance itself says how far through the cell the
    surface passes, so each cell contributes the fraction it actually fills.
    """
    low, high = part.bounds()
    span = high - low
    pitch = float(span.max()) / max(resolution, 4)
    dims = tuple(int(max(2, math.ceil(size / pitch))) for size in span)
    axes = [low[i] + (np.arange(dims[i]) + 0.5) * pitch for i in range(3)]
    points = np.stack(np.meshgrid(*axes, indexing="ij"), axis=-1)
    distance = part.sdf(points)
    filled = np.clip(0.5 - distance / pitch, 0.0, 1.0)
    volume = float(filled.sum()) * pitch ** 3
    inside = distance < 0
    if not inside.any():
        return 0.0, 0
    _, bodies = label_regions(inside)
    return volume, bodies


def inspect_part(
    part, resolution: int = 32
) -> tuple[list[ModelIssue], Optional[float], int]:
    """Every check, plus the measurements the geometry check already took.

    A sweep asks the same question a few dozen times over, and the volume it
    wants to chart is a by-product of the check it has just run. Handing it
    back costs nothing and halves the work.
    """
    issues = check_equations(part)
    if issues:
        return issues, None, 0

    values = part.resolved_parameters()
    issues.extend(check_features(part, values))
    if issues:
        return issues, None, 0

    if not part.active_features():
        return [ModelIssue("error", "geometry", "part", "has no features left to build")], None, 0

    try:
        volume, bodies = measure_solid(part, resolution)
    except Exception as error:               # noqa: BLE001
        return (
            [ModelIssue("error", "geometry", "part", f"cannot be evaluated: {error}")],
            None,
            0,
        )

    if volume <= 0:
        issues.append(
            ModelIssue("error", "geometry", "part", "builds to nothing at this size")
        )
        return issues, volume, bodies
    if bodies > 1:
        issues.append(
            ModelIssue(
                "warning", "geometry", "part",
                f"has fallen into {bodies} separate bodies",
            )
        )
    return issues, volume, bodies


def check_part(part, geometry: bool = True, resolution: int = 32) -> list[ModelIssue]:
    """Every check, in the order that makes the first failure the useful one."""
    if not geometry:
        issues = check_equations(part)
        if issues:
            return issues
        return check_features(part, part.resolved_parameters())
    return inspect_part(part, resolution)[0]


def is_buildable(part, resolution: int = 32) -> bool:
    return not any(issue.severity == "error" for issue in check_part(part, True, resolution))
